src/mpid/ch4/shm/posix/posix_win.h

```diff
@@ -384,5 +384,139 @@ static inline int MPIDI_POSIX_mpi_win_lock_all(int assert, MPIR_Win * win)
+                                                0 /* symmetric_flag */ ,
+                                                &posix_win->shm_mutex_segment_handle,
+                                                (void **) &posix_win->shm_mutex_ptr);
+
+    if (win->comm_ptr->rank == 0)
+        MPIDI_POSIX_rma_mutex_init(posix_win->shm_mutex_ptr);
+
+    /* No barrier is needed here, because the CH4 generic routine does it */
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_POSIX_MPI_WIN_ALLOCATE_SHARED_HOOK);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+#undef FUNCNAME
+#define FUNCNAME MPIDI_POSIX_mpi_win_create_dynamic_hook
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_mpi_win_create_dynamic_hook(MPIR_Win * win)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_POSIX_MPI_WIN_CREATE_DYNAMIC_HOOK);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_POSIX_MPI_WIN_CREATE_DYNAMIC_HOOK);
+
+    MPIDI_POSIX_win_t *posix_win ATTRIBUTE((unused)) = &win->dev.shm.posix;
+    posix_win->shm_mutex_ptr = NULL;
+
+    /* No optimization */
+
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_POSIX_MPI_WIN_CREATE_DYNAMIC_HOOK);
+    return mpi_errno;
+}
+
+MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_mpi_win_attach_hook(MPIR_Win * win ATTRIBUTE((unused)),
+                                                             void *base ATTRIBUTE((unused)),
+                                                             MPI_Aint size ATTRIBUTE((unused)))
+{
+    /* No optimization */
+    return MPI_SUCCESS;
+}
+
+MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_mpi_win_detach_hook(MPIR_Win * win ATTRIBUTE((unused)),
+                                                             const void *base ATTRIBUTE((unused)))
+{
+    /* No optimization */
+    return MPI_SUCCESS;
+}
+
+#undef FUNCNAME
+#define FUNCNAME MPIDI_POSIX_mpi_win_free_hook
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_mpi_win_free_hook(MPIR_Win * win)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_POSIX_MPI_WIN_FREE_HOOK);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_POSIX_MPI_WIN_FREE_HOOK);
+
+    if (MPIDI_CH4U_WIN(win, shm_allocated) == 1) {
+        MPIDI_POSIX_win_t *posix_win = &win->dev.shm.posix;
+        MPIR_Assert(posix_win->shm_mutex_ptr != NULL);
+
+        /* destroy and detach shared mutex */
+        if (win->comm_ptr->rank == 0)
+            MPIDI_POSIX_rma_mutex_destroy(posix_win->shm_mutex_ptr);
+
+        mpi_errno = MPL_shm_seg_detach(posix_win->shm_mutex_segment_handle,
+                                       (char **) &posix_win->shm_mutex_ptr,
+                                       sizeof(MPL_proc_mutex_t));
+        if (mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+
+        mpi_errno = MPL_shm_hnd_finalize(&posix_win->shm_mutex_segment_handle);
```

Comment:
I defined the generic routine `MPIDI_CH4U_allocate_shm_segment` just because of the complicated steps at allocation. But you are right that it is not symmetric. I will add `MPIDI_CH4U_destroy_shm_segment` in the next push.